addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/net/WaitingSessionListener.java

```diff
@@ -0,0 +1,63 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.russound.internal.net;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implementation of {@link SocketSessionListener} that allows a caller to wait for a response via
+ * {@link #getResponse()}
+ *
+ * @author Tim Roberts
+ */
+public class WaitingSessionListener implements SocketSessionListener {
+
+    /**
+     * Cache of responses that have occurred
+     */
+    private BlockingQueue<Object> _responses = new ArrayBlockingQueue<Object>(5);
+
+    /**
+     * Will return the next response from {@link #_responses}. If the response is an exception, that exception will
+     * be thrown instead.
+     *
+     * @return a non-null, possibly empty response
+     * @throws Exception an exception if one occurred during reading
+     */
+    public String getResponse() throws Exception {
+        final Object lastResponse = _responses.poll(5, TimeUnit.SECONDS);
+        if (lastResponse instanceof String) {
+            return (String) lastResponse;
+        } else if (lastResponse instanceof Exception) {
+            throw (Exception) lastResponse;
+        } else if (lastResponse == null) {
+            throw new Exception("Didn't receive response in time");
+        } else {
+            return lastResponse.toString();
+        }
+    }
+
+    @Override
+    public void responseReceived(String response) {
+        try {
+            _responses.put(response);
+        } catch (InterruptedException e) {
```

Comment:
Not in this case as they all deal with the thread ending.  However, I will rearrange the code a bit to make it more explicit